### ISSUE TITLE
Flow pragma does not inherit styles from comment

### ISSUE THREAD

**Issue by StreetStrider** [borela/naomi#148]
Hello, @borela!
I use special background styling for zones like comments, strings and regexes. So I catch this issue pretty easy.

![pragma-white](https://user-images.githubusercontent.com/2124761/41800118-c58f61a8-767c-11e8-9dc4-850fc8c33c24.png)

```
/* @flow */
/* ::
type $Glob$Options = Object
*/
```

The deal is that flow pragma got scope `source.js meta.naomi keyword.other.prolog.flow`, while other comments have
`source.js meta.naomi comment.block.flow punctuation.definition.comment.begin.flow`.
As you can see there's no scope of `comment` or `comment.block` for pragma, so it's not treated as comment (in my scheme it must be with blue bg in that particular case). I think pragma is just a special kind of comment, so it must have `comment.block.flow` somewhere in its scope.

I can hard-fix this on my end, but I think other schemes may win from this fix.

**Comment by borela** [comment]
Sorry for the delay, I agree with this fix, do you want to send the patch as a pull request?

**Comment by StreetStrider** [comment]
@borela no probs for the delay, I just file tickets so we can track them without forgetting.
I can definitely prepare patch, but I have zero or little experience with sublime-syntaxes, so I'll need to investigate into project (it must be easy, I presume).

**Comment by borela** [comment]
https://github.com/borela/naomi/blob/master/syntaxes/flow1/prolog.sublime-syntax This is the file that assigns the scope to the pragma, maybe the file itself must be renamed too because at the time I didn't know what to call it.

**Comment by borela** [comment]
There are just a few places that references it https://github.com/borela/naomi/search?q=flow1%2Fprolog&unscoped_q=flow1%2Fprolog&type=Code

**Comment by StreetStrider** [comment]
@borela can I use backrefs?
My current idea is:
```
        (?x)
        (?> (?<= \/\/ \s* ) @flow
          | (?<= \/\* \s* ) @flow (?= \s* \*\/ )
        )
```
for `flow1.prolog`. But it outputs error. Can you check my approach? The regexp itself seems to be correct, but I'm not sure for sublime regex engine. Need some help.

**Comment by borela** [comment]
No need in this case, i believe you want to apply different scopes to different parts of the pragma, in that case, you can use capturing groups:
```
    - match: |
        (?x)
        (?> \/\/\s*(@flow)
          | \/\*\s*(@flow)\s\*\/
        )
      captures:
        0: applies.to.the.entire.match
        1: this.scope.applies.to.first.capture 
        2: this.scope.applies.to.second.capture
      pop: true

**Comment by StreetStrider** [comment]
For now I'm starting to think of it at different angle. I think `@flow` must be the only thing that match, but it must be recognizable *only* in comment's context. So, outer scope will be comment automatically (because outer rule to match will be comment), and prolog will match for `@flow`. I don't know how to express it, but seemes it's possible, since Naomi already got wide support of syntax highlighting *inside* comments. I will try to learn on examples.

**Comment by StreetStrider** [comment]
Just to file. There's another place where comment context is not inherited:
![scope-comment](https://user-images.githubusercontent.com/2124761/42077428-8c627cda-7b81-11e8-9433-51a20d39b788.png)
When inside `/* :type */` (2) everything is ok. Inside `/* :: some flowtyped stuff */` (1) may have de-commented scope (`source.jsx meta.naomi meta.tag.other.html.jsx.jsx1 meta.jsx-fence.jsx.jsx1 punctuation.definition.tag.begin.jsx.jsx1` — no `comment.block.flow` present).

**Comment by borela** [comment]
Currently, the syntax is making a distinction between the declaration of a generic:

```
function abc<T, N>(){}
// and its usage:
abc<number, string>()
```
![image](https://user-images.githubusercontent.com/11317458/42094023-c6ee9fca-7b84-11e8-8225-23e4b0fab57d.png)

It's possible to highlight those cases by adding the line:

```
    - include: Packages/Naomi/syntaxes/flow1/generic/arguments.sublime-syntax
```

Before this line https://github.com/borela/naomi/blob/master/syntaxes/flow1/comment.sublime-syntax#L41

The only issue right now is that it would highlight template parameters as types:

![image](https://user-images.githubusercontent.com/11317458/42094105-1877c344-7b85-11e8-8576-9567d73b9398.png)

(You also caught a bug in function calls not expecting comments).

To continue highlighting generic variables, a special condition near function and class declarations will be needed.

**Comment by borela** [comment]
> I understood why you draw attention to this distinction, but thinking further — in fact they are types. It's a type variable, so it is a type. It's like function parameters, which is a locals inside function, but, more specifically they come from function signature, not var decls. Maybe they can be coded accordingly (like, for a raw example type is a scope for all variables type.arg for generic arguments).

That makes sense and just highlighting them as types would simplify the syntax a bit.

**Comment by borela** [comment]
I sent some patchs on the master branch which should solve most issues:

![image](https://user-images.githubusercontent.com/11317458/42095797-07ae8cbe-7b8a-11e8-9305-f1532627d9bc.png)

Please check if it's working.

**Comment by StreetStrider** [comment]
@borela hello. Generic normalization went ok in my case.
![_2018-07-03_10-48-52](https://user-images.githubusercontent.com/2124761/42206158-ffed6e46-7eae-11e8-9cfb-4119b72ce02e.png)

**Comment by borela** [comment]
Thank you for confirming it :D I am leaving this open because we still have the issue of @flow not being highlighted in block comments other than `// @flow` or `/* @flow */`.

**Comment by StreetStrider** [comment]
@borela agreed. :+1: I'l try to learn something about syntaxes and try to fix this by myself. My idea is still the same, flow prolog must be recognizable inside comment context, so both scopes will apply accordingly. Will try to investigate this way.
